File: letters/generate-letters.js

const fs = require('fs');
const exec = require('child_process').exec;

const schools = [
    ['brown', 'a note'],
    ['blue', 'a note'],
    ['purple', 'a note']
]
// read the main file
const mainfile = fs.readFileSync('../tex/template.tex', 'utf8');
// read the template part
const sampleforallFile = fs.readFileSync('../tex/sameforall.tex', 'utf8');


schools.forEach((school, index) => {

    // read the school configuration file
    let parameterFile = '../school-customization/' + school[0] + '.tex'
    let parameters = fs.readFileSync(parameterFile, 'utf8');

    // decode the configuration file
    let schoolinfo = parameters.split('%---').filter(l => l != '')
    
    let nextFile = JSON.parse(JSON.stringify(mainfile));

    // add the text color command
    currAll = nextFile.replace('\\input{../tex/sameforall.tex}', sampleforallFile)

    // look into each school
    schoolinfo.forEach(entry => {
        let entryLines = entry.trim().split('\n')

        if (entryLines.length > 1) {
            // console.log(entryLines)

            let content = entryLines[1].trim()

            // replace the template file with the info in the configuration file
            currAll = currAll.replace(entryLines[0], content)


        }
    })

    // output the school tex
    let name = school[0].toUpperCase()
    let tex_file = name + '-letter.tex'
    let pdf_file = name + '-letter.pdf'

    fs.writeFile(tex_file, currAll, () => { console.log(school) })


    // compile the school text into PDF
    exec('xelatex --output-directory=tmps ' + tex_file,
        function (error, stdout, stderr) {
            console.log('stdout: ' + stdout);
            console.log('stderr: ' + stderr);
            if (error !== null) {
                console.log('exec error: ' + error);
            } else {
                // move the PDF into another folder
                exec('mv tmps/' + pdf_file + ' pdfs/' + pdf_file,
                    function (error, stdout, stderr) {
                        console.log('stdout: ' + stdout);
                        console.log('stderr: ' + stderr);
                        if (error !== null) {
                            console.log('exec error: ' + error);
                        } else {
                            console.log('total : ' + schools.length)
                        }
                    });
            }
        });

})

File: letters/generate-letters.py
import subprocess
import os

schools = [
    ['brown', 'a note'],
    ['blue', 'a note'],
    ['purple', 'a note']
]

# Read the main file
with open('../tex/template.tex', 'r', encoding='utf-8') as file:
    mainfile = file.read()

# Read the template part
with open('../tex/sameforall.tex', 'r', encoding='utf-8') as file:
    sampleforallFile = file.read()

for school in schools:
    # Read the school configuration file
    parameter_file = '../school-customization/' + school[0] + '.tex'
    with open(parameter_file, 'r', encoding='utf-8') as file:
        parameters = file.read()

    # Decode the configuration file
    schoolinfo = [line for line in parameters.split('%---') if line != '']

    next_file = mainfile[:]
    # Add the text color command
    curr_all = next_file.replace('\\input{../tex/sameforall.tex}', sampleforallFile)

    # Look into each school
    for entry in schoolinfo:
        entry_lines = entry.strip().split('\n')
        if len(entry_lines) > 1:
            content = entry_lines[1].strip()
            # Replace the template file with the info in the configuration file
            curr_all = curr_all.replace(entry_lines[0], content)

    # Output the school tex
    name = school[0].upper()
    tex_file = name + '-letter.tex'
    pdf_file = name + '-letter.pdf'

    with open(tex_file, 'w', encoding='utf-8') as file:
        file.write(curr_all)
    print(school)

    # Compile the school text into PDF
    try:
        subprocess.run(f'xelatex --output-directory=tmps {tex_file}', check=True, shell=True, text=True)
        # Move the PDF into another folder
        subprocess.run(f'mv tmps/{pdf_file} pdfs/{pdf_file}', check=True, shell=True, text=True)
    except subprocess.CalledProcessError as e:
        print(f'An error occurred: {e}')
    else:
        print('Total:', len(schools))
